### PR TITLE
Implement `Coordinator`, a reaper to spawn the pivot executable

### PR DESCRIPTION
**Pull request by emostov**
This PR introduces the `Coordinator` which waits for both the pivot executable and quorum secret to be written to the file system. Once those are both detected, the coordinator runs the pivot executable.

With the current protocol API, the current way to do this is `Load` the executable and `Provision` the quorum key by posting shards and sending the `Reconstruct` command. This flow is demonstrated in in the `coordinator_e2e` integration test.

The biggest design question is: https://github.com/tkhq/qos/pull/25/files#r880067438

### CODE REVIEW

File: qos-core/src/cli.rs [hunk 1]
```diff
@@ -12,14 +13,23 @@ pub struct EnclaveOptions {
...
 	pub fn new() -> Self {
-		Self { cid: None, port: None, usock: None, mock: false }
+		Self {
+			cid: None,
+			port: None,
+			usock: None,
+			mock: false,
+			secret_file: SECRET_FILE.to_owned(),
+			pivot_file: PIVOT_FILE.to_owned(),
+		}
```
Comment (emostov) [inline qos-core/src/cli.rs:29]: To make multi threaded testing practical, I changed both the secret and pivot file to be injectable. This way we can ensure tests running in parallel are not trying to access the same files. 

Additionally, it helps use decouple our architecture from using global variables - which should make future refactors easier

File: qos-core/src/cli.rs [hunk 2]
```diff
@@ -12,14 +13,23 @@ pub struct EnclaveOptions {
...
 impl EnclaveOptions {
 	pub fn new() -> Self {
-		Self { cid: None, port: None, usock: None, mock: false }
+		Self {
+			cid: None,
+			port: None,
+			usock: None,
+			mock: false,
+			secret_file: SECRET_FILE.to_owned(),
+			pivot_file: PIVOT_FILE.to_owned(),
```
Comment (emostov) [inline qos-core/src/cli.rs:28]: We default to the old values so they don't need to be specified at the command line. Alternately we could make these options and not have defaults

File: qos-core/src/cli.rs [hunk 3]
```diff
@@ -111,35 +151,68 @@ impl CLI {
...
-		SocketServer::listen(addr, executor).unwrap();
+		Coordinator::execute(options);
 	}
 }
 
 #[cfg(test)]
 mod test {
 	use super::*;
 
+	// TODO: add tests for parsing file paths - verify that file paths are valid
```
Comment (emostov) [inline qos-core/src/cli.rs:164]: Make an issue for this

File: qos-core/src/coordinator.rs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+				println!("Pivot executable exited successfully ...");
+				break
+			} else {
+				println!(
+					"Re-spawning pivot executable child process - {}",
+					status
+				);
+				child_process =
+					pivot.spawn().expect("Process failed to execute ...");
+			}
```
Comment (emostov) [inline qos-core/src/coordinator.rs]: I actually don't think we should re-spawn here - probably should instead just exit. Any suggestions @jack-kearney ?

File: qos-core/src/lib.rs [hunk 5]
```diff
@@ -4,6 +4,17 @@ pub mod cli;
...
 pub mod protocol;
 pub mod server;
+
+#[cfg(not(feature = "vm"))]
+pub const SECRET_FILE: &str = "./qos.secret";
+#[cfg(feature = "vm")]
+pub const SECRET_FILE: &str = "/qos.secret";
+
+#[cfg(not(feature = "vm"))]
+pub const PIVOT_FILE: &str = "../target/debug/pivot_ok";
```
Comment (emostov) [inline qos-core/src/lib.rs:18]: This comes from the binary in _qos-test/src/bin/pivot_ok.rs_

File: qos-test/src/bin/host_cli.rs [hunk 6]
```diff
@@ -0,0 +1,4 @@
+#[tokio::main]
+async fn main() {
+	qos_host::cli::CLI::execute().await;
+}
```
Comment (emostov) [inline qos-test/src/bin/host_cli.rs:4]: These are hacky - but wasn't sure how to ensure that the executables from other crates built just with `cargo test`, so instead just made new executables that are simply the `CLI`s

File: qos-test/tests/coordinator.rs [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+
+	// **Load the executable**
+
+	// -- Convert the executable to bytes
+	let pivot_bytes = std::fs::read(PIVOT_OK_PATH).unwrap();
+
+	// -- Send that executable via the ProtocolLoad message
+	let load_msg = ProtocolMsg::LoadRequest(Load {
+		executable: pivot_bytes,
+		signatures: vec![],
```
Comment (emostov) [inline qos-test/tests/coordinator.rs:68]: Leaving the signatures out for now since this logic will likely change a bit

Comment (emostov) [reply qos-test/tests/coordinator.rs:68]: But I am guessing we probably want something like `Vec<(Signature, PubKey)>`

File: qos-test/tests/provision_e2e.rs [hunk 8]
```diff
@@ -20,10 +20,16 @@ async fn end_to_end() {
 	let health_url = format!("{}/{}", url, "health");
 	let message_url = format!("{}/{}", url, "message");
 
+	let pivot_file = "./end-to-end.pivot".to_string();
+	let secret_file = "./end-to-end.secret".to_string();
+
 	// Spawn enclave
+	let pivot_file2 = pivot_file.clone();
+	let secret_file2 = secret_file.clone();
```
Comment (emostov) [inline qos-test/tests/provision_e2e.rs]: We just need to clone so we can "move" them to a new thread.

File: qos-core/src/coordinator.rs (hunk 4 above)
Comment (emostov) [reply qos-core/src/coordinator.rs]: I changed it to never respawn - see: [25b86e8](https://github.com/tkhq/qos/pull/25/commits/25b86e87b16a2a35f0f706c410eec0b4e6275634)

File: qos-core/src/coordinator.rs [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+		std::thread::spawn(move || {
+			let executor = Executor::new(
+				opts.nsm(),
+				opts.secret_file(),
+				opts.pivot_file(),
+			);
+			SocketServer::listen(opts.addr(), executor).unwrap();
+		});
+
```
Comment (emostov) [inline qos-core/src/coordinator.rs:26]: We could consider putting an outer loop here - as it stands the enclave (Coordinator) will just shut down once the loaded pivot executable exits. The main issue is that we will just try re-running the same pivot file again, which I am not sure if we want to do that

File: qos-core/src/cli.rs (hunk 1 above)
Comment (jack-kearney) [reply qos-core/src/cli.rs:29]: Great, makes a ton of sense to me

File: qos-core/src/coordinator.rs [hunk 10]
```diff
@@ -0,0 +1,64 @@
+//! The coordinator is responsible for initializing the enclave's primary
+//! processes. Concretely it spawns the enclave server and launches the "pivot"
+//! executable once it becomes available in the file system.
+//!
+//! The pivot is an executable the enclave runs to initialize the secure
```
Comment (jack-kearney) [inline qos-core/src/coordinator.rs:5]: NIT: Aren't the "pivot" the "secure application" the same thing? A little bit confused by this comment

File: qos-core/src/coordinator.rs (hunk 9 above)
Comment (jack-kearney) [reply qos-core/src/coordinator.rs:26]: Yeah, this is an interesting consideration. I think we'll typically want to restart the secure application but there are definitely legitimate use-cases for not wanting to do that: Imagine QOS being used to execute a single task -- I think we'll almost certainly want to build that functionality at some point soon.

For right now, let's just leave this as is but I expect that we should make this configurable in the very near future.

File: qos-core/src/lib.rs (hunk 5 above)
Comment (jack-kearney) [reply qos-core/src/lib.rs:18]: This feels kind of wonky -- can you elaborate on why we need it?

File: qos-test/src/bin/host_cli.rs (hunk 6 above)
Comment (jack-kearney) [reply qos-test/src/bin/host_cli.rs:4]: Don't think I follow. Can you elaborate?

File: qos-test/tests/coordinator.rs (hunk 7 above)
Comment (jack-kearney) [reply qos-test/tests/coordinator.rs:68]: Yeah agreed that we should leave this out for now

File: qos-test/tests/coordinator.rs [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+	protocol::{Load, ProtocolMsg, Provision},
+};
+use qos_crypto::shares_generate;
+
+const PIVOT_OK_PATH: &str = "../target/debug/pivot_ok";
+const PIVOT_ABORT_PATH: &str = "../target/debug/pivot_abort";
+const PIVOT_PANIC_PATH: &str = "../target/debug/pivot_panic";
+
+#[tokio::test]
+async fn coordinator_e2e() {
```
Comment (jack-kearney) [inline qos-test/tests/coordinator.rs:15]: Shouldn't this first take the shares and provision the device and only afterwards load the executable and pivot? I think we should actually add some validation into the coordinator to ensure that the machine has already been provisioned.

**Pull request merged**